### PR TITLE
Issue #889772 by tuutti, stefan.r, opdavies, Sutharsan, Perignon, pjc…

### PR DESCRIPTION
**Pull request by jenlampton**
PR for https://github.com/backdrop/backdrop-issues/issues/2013

### CODE REVIEW

File: core/modules/user/user.pages.inc [hunk 1]
```diff
@@ -113,22 +119,34 @@ function user_pass_reset($uid, $timestamp, $hashed_pass, $immediate_login = FALS
...
+    // current page so the password reset process can continue.
     if ($user->uid == $uid) {
-      backdrop_set_message(t('You are logged in as %user. <a href="!user_edit">Change your password.</a>', array('%user' => $user->name, '!user_edit' => url("user/$user->uid/edit"))));
+      // Preserve the current destination (if any) and ensure the redirect goes
+      // back to the current page; any custom destination set in
+      // hook_user_logout() and intended for regular logouts would not be
+      // appropriate here.
+      $destination = array();
+      if (isset($_GET['destination'])) {
+        $destination = drupal_get_destination();
```
Comment (klonos) [inline core/modules/user/user.pages.inc]: https://api.backdropcms.org/api/backdrop/core%21includes%21common.inc/function/backdrop_get_destination/1

File: core/modules/user/user.pages.inc [hunk 2]
```diff
@@ -213,6 +231,14 @@ function user_pass_reset_form_submit($form, &$form_state) {
  * Menu callback; logs the current user out, and redirects to the home page.
  */
 function user_logout() {
+  user_logout_current_user();
+  drupal_goto();
```
Comment (klonos) [inline core/modules/user/user.pages.inc]: `backdrop_goto();`

File: core/modules/simpletest/tests/simpletest.test [hunk 3]
```diff
@@ -362,6 +369,46 @@ EOF;
     $urls = $this->xpath('//a[text()=:text]', array(':text' => 'A second "even more weird" link, in memory of George O\'Malley'));
     $this->assertEqual($urls[0]['href'], 'link2', 'Match with mixed single and double quotes.');
   }
+
+  /**
+   * Tests that cookies set during a request are available for testing.
+   */
+  public function testCookies() {
+    // Check that the $this->cookies property is populated when a user logs in.
+    $user = $this->drupalCreateUser();
```
Comment (klonos) [inline core/modules/simpletest/tests/simpletest.test]: `$this->backdropCreateUser();`

File: core/modules/simpletest/tests/simpletest.test [hunk 4]
```diff
@@ -362,6 +369,46 @@ EOF;
...
+    $edit = array('name' => $user->name, 'pass' => $user->pass_raw);
+    $this->drupalPost('<front>', $edit, t('Log in'));
```
Comment (klonos) [inline core/modules/simpletest/tests/simpletest.test]: `$this->backdropPost`

File: core/modules/simpletest/tests/simpletest.test [hunk 5]
```diff
@@ -362,6 +369,46 @@ EOF;
...
+    $this->assertEqual(count($this->cookies), 1, 'A cookie is set when the user logs in.');
+
+    // Check that the name and value of the cookie match the request data.
+    $cookie_header = $this->drupalGetHeader('set-cookie', TRUE);
```
Comment (klonos) [inline core/modules/simpletest/tests/simpletest.test]: `$this->backdropGetHeader`

**Comment by jenlampton** [comment]
Fixed and pushed up a new PR

**Comment by jenlampton** [comment]
Rebased and pushed a new PR. since test fail links here are missing.

**Comment by jenlampton** [comment]
Test runner failed. closing and opening.

**Comment by quicksketch** [comment]
Looks like tests are failing because of a critical error:

```
PHP Fatal error: Cannot redeclare BackdropWebTestCase::$cookies in /home/test/github/pr/1887/core/modules/simpletest/backdrop_web_test_case.php on line 974
```

**Comment by quicksketch** [comment]
Re-pushed a fix to the last commit to correct the double defining of the `$cookies` variable that was causing the PHP fatal.

**Comment by jenlampton** [comment]
Rebased and pushed again.

**Comment by ghost** [comment]
@jenlampton Can you rebase this PR please? (assuming it's still needed) It's the last PR from https://github.com/backdrop/backdrop-issues/issues/2013 (it was missed when that issue was closed).

**Comment by klonos** [comment]
Replaced by the following individual PRs:
- https://github.com/backdrop/backdrop/pull/4460
- https://github.com/backdrop/backdrop/pull/4461
- https://github.com/backdrop/backdrop/pull/4462

**Pull request closed without merging**